mcs/class/System/System.Net.NetworkInformation/IPInterfaceProperties.cs

```diff
@@ -393,28 +390,17 @@ static IPAddressInformationCollection Win32FromAnycast (IntPtr ptr)
 			get {
 				var col = new GatewayIPAddressInformationCollection ();
 				try {
-					Win32_IP_ADAPTER_INFO v4info = Win32NetworkInterface2.GetAdapterInfoByIndex (mib4.Index);
 					// FIXME: should ipv6 DhcpServer be considered?
-
-					var a = v4info.GatewayList;
-					if (!String.IsNullOrEmpty (a.IpAddress)) {
-						col.InternalAdd(new SystemGatewayIPAddressInformation(IPAddress.Parse (a.IpAddress)));
-						AddSubsequently (a.Next, col);
+					Win32_IP_ADAPTER_GATEWAY_ADDRESS a;
+					for (IntPtr p = addr.FirstGatewayAddress; p != IntPtr.Zero; p = a.Next) {
+						a = (Win32_IP_ADAPTER_GATEWAY_ADDRESS) Marshal.PtrToStructure (p, typeof (Win32_IP_ADAPTER_GATEWAY_ADDRESS));
```

Comment:
Maybe do some sanity check that we at least have the expected struct length (included as a field in the unmanaged struct)?

Reply:
Same comment above would apply to this function as well

Reply:
OK, same comment as above.